### PR TITLE
Test against localstack

### PR DESCRIPTION
**Pull request by markreesmoj**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -3,12 +3,38 @@ version: 2.1
 orbs:
   hmpps: ministryofjustice/hmpps@3.0
   owasp: entur/owasp@0.0.10
+executors:
+  validator-integration:
+    docker:
+      - image: circleci/openjdk:11-jdk
+      - image: localstack/localstack:0.11.2
+        environment:
```
Comment (nickmcmahon01) [inline .circleci/config.yml:11]: I've never checked if we actually need all this, or is it just magic cruft that we copy around.

Comment (markreesmoj) [reply .circleci/config.yml:11]: I will try taking bits out on a separate branch and see what happens

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstier/config/AwsLocalStackConfiguration.kt [hunk 2]
```diff
@@ -33,14 +34,24 @@ class AwsLocalStackConfiguration(
...
+
   @Primary
   @Bean(name = ["hmppsDomainEvents"])
   fun hmppsDomainEventsAmazonSNSAsync(
     @Value("\${hmpps-domain-events.sns-endpoint}") serviceEndpoint: String
   ): AmazonSNSAsync {
     return AmazonSNSAsyncClientBuilder.standard()
       .withEndpointConfiguration(AwsClientBuilder.EndpointConfiguration(serviceEndpoint, region))
-      .withCredentials(AWSStaticCredentialsProvider(AnonymousAWSCredentials()))
+      .withCredentials(AWSStaticCredentialsProvider(BasicAWSCredentials("key", "secret")))
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstier/config/AwsLocalStackConfiguration.kt:54]: Half a day's work to figure this out

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstier/integration/JwtAuthHelper.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
+package uk.gov.justice.digital.hmpps.hmppstier.integration
```
Comment (markreesmoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstier/integration/JwtAuthHelper.kt:1]: borrowed from assessments-api

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstier/integration/MockedEndpointsTestBase.kt [hunk 4]
```diff
@@ -29,16 +40,58 @@ import uk.gov.justice.digital.hmpps.hmppstier.integration.ApiResponses.restricti
 import uk.gov.justice.digital.hmpps.hmppstier.integration.ApiResponses.restrictiveRequirementsResponse
 import java.nio.file.Files
 import java.nio.file.Paths
+import java.time.Duration
 import java.time.LocalDate
+import java.util.UUID
 
 abstract class MockedEndpointsTestBase : IntegrationTestBase() {
+
```
Comment (markreesmoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstier/integration/MockedEndpointsTestBase.kt:48]: borrowed from assessments-api

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstier/integration/OAuthMockServer.kt [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package uk.gov.justice.digital.hmpps.hmppstier.integration
```
Comment (markreesmoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstier/integration/OAuthMockServer.kt:1]: borrowed from assessments-api

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstier/integration/JwtAuthHelper.kt (hunk 3 above)
Comment (nickmcmahon01) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstier/integration/JwtAuthHelper.kt:1]: https://github.com/search?q=org%3Aministryofjustice+JwtAuthHelper&type=code

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstier/config/AwsLocalStackConfiguration.kt [hunk 6]
```diff
@@ -33,14 +34,24 @@ class AwsLocalStackConfiguration(
       .build()
   }
 
+  @Bean
+  fun calculationCompleteEventsAmazonSQSAsync(
+    @Value("\${calculation-complete.sqs-endpoint}") serviceEndpoint: String
+  ): AmazonSQSAsync {
+    return AmazonSQSAsyncClientBuilder.standard()
+      .withEndpointConfiguration(AwsClientBuilder.EndpointConfiguration(serviceEndpoint, region))
+      .withCredentials(AWSStaticCredentialsProvider(AnonymousAWSCredentials()))
```
Comment (nickmcmahon01) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstier/config/AwsLocalStackConfiguration.kt:43]: Why does this one work? is it pub/sub issue or an sqs/sns issue?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstier/integration/MandateForChangeTest.kt [hunk 7]
```diff
@@ -27,11 +15,8 @@ class MandateForChangeTest : MockedEndpointsTestBase() {
     setupRestrictiveRequirements(crn)
     setupMaleOffenderWithRegistrations(crn)
 
-    listener.listen(calculationMessage(crn))
-    val tier = repo.findFirstByCrnOrderByCreatedDesc(crn)
-
-    Assertions.assertThat(tier?.data?.change?.tier).isEqualTo(ChangeLevel.ZERO)
-    Assertions.assertThat(tier?.data?.protect?.tier).isEqualTo(ProtectLevel.A)
+    calculateTierFor(crn)
+    expectTierCalculation("A0")
```
Comment (nickmcmahon01) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstier/integration/MandateForChangeTest.kt:19]: This looks much nicer, I'm glad we don't have a requirement to display them separately.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstier/integration/MockedEndpointsTestBase.kt [hunk 8]
```diff
@@ -231,5 +289,50 @@ abstract class MockedEndpointsTestBase : IntegrationTestBase() {
     )
   }
 
+  fun calculateTierFor(crn: String) {
```
Comment (nickmcmahon01) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstier/integration/MockedEndpointsTestBase.kt:292]: This has taken a lot of code out of the other tests, which has made them much clearer.

**Review by nickmcmahon01 — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstier/config/AwsLocalStackConfiguration.kt (hunk 6 above)
Comment (markreesmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstier/config/AwsLocalStackConfiguration.kt:43]: I think we publish to both SQS and SNS in these tests - SNS publish on localstack I think is what caused the error. Other applications have no credentials set here at all e.g. 

https://github.com/ministryofjustice/court-register/blob/76cc4b748d9faf3b6a610d1c895962d863cb7bf7/src/main/kotlin/uk/gov/justice/digital/hmpps/courtregister/config/LocalstackConfig.kt

I couldn't work out where the credentials come from there, maybe environment variables

 I can't find a reference for the problem now though

**Pull request merged**